PowerPC: multiple VLE-mode fixes

esaulenka opened the pull request:
I am tested Ghidra with binary for modern NXP MCP57xx MCU and I found a number of issues.

- XOR in VLE mode is valid. It is stated in https://www.nxp.com/files-static/32bit/doc/ref_manual/EREF_RM.pdf
> Instructions from primary opcode 4 and 31 in the standard instruction set decode the same for the
VLE instruction set, making all the instructions in those two primary opcodes available in VLE
mode

- ISEL works incorrectly when rD and rA points to the same register. In that case first step in semantic sections kills rA content.

- efscmpeq/efscmpgt/efscmplt modifies wrong bit. It is stated in https://www.nxp.com/files-static/32bit/doc/ref_manual/SPEPEM.pdf page 114
Also it can be tested with this code:
```
                     float  __stdcall  floatMIN (float  param_1 , float  param_2 )
                       assume vle = 0x1
     float             r3:4           <RETURN>
     float             r3:4           param_1
     float             r4:4           param_2
                     floatMIN
xxxxxx58 10  83  22  cc    efscmpgt   cr1 ,param_1 ,param_2
xxxxxx5c 7a  15  00  06    e_bgt      LAB_xxxxxx62
xxxxxx60 00  04           se_blr
                     LAB_xxxxxx62
xxxxxx62 01  43           se_mr      param_1 ,param_2
xxxxxx64 00  04           se_blr
```

- e_rlwinm with `if(...) goto ...` produces annoying warnings about unreachable code. Condition in `if` statement always constant (since it is a part of instruction itself), so better to write it in another form.

- modern NXP processors have extended signal processing instructions. I am added ones, that listed in https://www.nxp.com/docs/en/engineering-bulletin/EB689.pdf (note, it contain a lot of copy-paste errors). I don't know, maybe better to do another 'language' and add them there.

GhidorahRex commented:
I think the extended signal processing instructions should be placed in a separate .sinc file and possibly added as a separate PPC language variant. Although there's a lot of PPC variants already.

gregjhogan commented:
efststeq/efststgt/efststlt has the same issues as efscmpeq/efscmpgt/efscmplt see https://github.com/NationalSecurityAgency/ghidra/pull/2528

GhidorahRex commented on Ghidra/Processors/PowerPC/data/languages/Scalar_SPFP.sinc:
```diff
@@ -470,3 +467,97 @@ define pcodeop ConvertFloatingPointFromUnsignedFraction;
 {
   CRFD = A:4 f< B:4;
 }
+
```
I think these are probably better moved to SPE_EFV.sinc

GhidorahRex replied:
I should clarify. This will not build as is because the vector instructions rely on the registers being greater than 4 bytes in length. Our 32-bit PPC processors will fail to build with this included here. Moved to SPE_EFV.sinc will resolve the issue.

GhidorahRex commented on Ghidra/Processors/PowerPC/data/languages/ppc_vle.sinc:
```diff
@@ -800,36 +800,34 @@ IMM16B: val						is IMM_0_10_VLE & IMM_16_20_VLE [ val = (IMM_16_20_VLE << 11) |
+	tmpM2 = tmpM2 << (31-MBL);
+	tmpM2 = ~tmpM2;
+
+	tmpM = zext(ME:1 >= MBL:1)*tmpM1 + zext(ME:1 < MBL:1)*tmpM2;
```
Nice use of the choosery pattern here

esaulenka commented:
> I should clarify. This will not build as is because the vector instructions rely on the registers being greater than 4 bytes in length. Our 32-bit PPC processors will fail to build with this included here. Moved to SPE_EFV.sinc will resolve the issue.

Sorry for delay and thanks for noticing.
As I understood these new instructions supported only by VLE-version of CPUs. So I added it to a new file that should be used for `PowerPC:BE:64:VLE-32addr`. I don't want to add a new CPU, because even current set a bit confusing for ones who don't know the whole history of PowerPC cores.

GhidrAuto commented:
Hey esulanke, sorry to bug you, but I have a dumb question followed by slightly less dumb questions (I HOPE). are these commits in the latest build, 11.1.2? or do I need to learn how to build out this file https://github.com/esaulenka/ghidra/tree/ppc_fixes?  with the gradle? forgive my terminology if it is incorrect. 

the issue I am Having. with a powerPC VLE File is(I Will include the processor type when I get the info back from the sender of the file); 
  I chose the Power ISA 3.0 Big endian w/VLE,EVX and 32-bit addressing for this file I am working on. (altivec didn't work nearly at all) the issue I am dealing with is missing efsMIN, efsMAX, the broken spaces of code below is where they belong. Also a large amount of "WARNING: Bad instruction - Truncating control flow here *", and "-- Flow Override: CALL_RETURN (CALL_TERMINATOR)". I am able to work through the broken functions and make some educated guesses parameters that I need to manipulate, however I would really like to clean this up. Any and all suggestions welcomed and appreciated. I am willing to put considerable time into fixing this or assisting with a fix if someone can point me in the right direction.

        090bccca 70 07 e7 80     e_lis      r0,0x3f80
        090bccce 13              ??         13h                        
        090bcccf c7              ??         C7h
        090bccd0 02 b0           se_mtar    r8,r27
        090bccd2 10              ??         10h
        090bccd3 e5              ??         E5h
        090bccd4 f2              ??         F2h
        090bccd5 b1              ??         B1h
                             LAB_090bccd6                                    XREF[1]:     090bcc72(j)  
        090bccd6 10              ??         10h
        090bccd7 06              ??         06h
        090bccd8 f2              ??         F2h
        090bccd9 b1              ??         B1h
        090bccda d1 4f           se_stw     r4,0x4(r31)
        090bccdc 70 8a e0 80     e_lis      r4,0x5080
        090bcce0 10              ??         10h
        090bcce1 e7              ??         E7h
        090bcce2 ea              ??         EAh
        090bcce3 b0              ??         B0h
        090bcce4 50 84 02 e0     e_lwz      r4,0x2e0(r4)
        090bcce8 73 e8 e0 02     e_lis      r31,0x4002
        090bccec 10              ??         10h
        090bcced 00              ??         00h
        090bccee ea              ??         EAh
        090bccef b0              ??         B0h
        090bccf0 1f ff 8a 04     e_add16i   r31,r31,-0x75fc
        090bccf4 70 61 e1 26     e_lis      r3,0x926
        090bccf8 d0 7f           se_stw     r7,0x0(r31)=>DAT_ffff8a04                        = ??
        090bccfa 1c 63 5d 78     e_add16i   r3=>DAT_09265d78,r3,0x5d78
        090bccfe d1 0f           se_stw     r0,0x4(r31)=>DAT_ffff8a08                        = ??
        090bcd00 79 ef 66 31     e_bl       2D_08fb3330                                      undefined 2D_08fb3330()
                             -- Flow Override: CALL_RETURN (CALL_TERMINATOR)
        090bcd04 70 e1 e1 32     e_lis      r7,0x932
        090bcd08 50 e7 7a 5c     e_lwz      r7=>DAT_40a00000,offset PTR_DAT_09327a5c(r7)     = ??
                                                                                             = 40a00000
        090bcd0c 10 63 3a c9     efsdiv     r3,r3,r7
        090bcd10 79 ef 5b d9     e_bl       FUN_08fb28e8                                     undefined FUN_08fb28e8()
                             -- Flow Override: CALL_RETURN (CALL_TERMINATOR)
        090bcd14 70 e1 e1 32     e_lis      r7,0x932
        090bcd18 50 e7 7a 5c     e_lwz      r7,offset PTR_DAT_09327a5c(r7)                   = 40a00000
        090bcd1c 10 03 3a c8     efsmul     r0,r3,r7
        090bcd20 70 e1 e1 32     e_lis      r7,0x932
        090bcd24 50 e7 7a 6c     e_lwz      r7,offset DAT_09327a6c(r7)                       = 43FA0000h
        090bcd28 10              ??         10h
        090bcd29 00              ??         00h
        090bcd2a 3a b1 70 e1     e_lha      r21,0x70e1(r17)
        090bcd2e e1 32           se_ble     cr0,LAB_090bcd92
        090bcd30 50 e7 7a 80     e_lwz      r7,0x7a80(r7)

esaulenka commented:
hi @GhidrAuto 

I made a quick research. Looks like the first missting opcodes are `maddhd` and `maddhdu`. It's specified in `ppc_isa.sinc` and haven't touched in my fix.
I think would be better to open a new ticket. Sample binary and exact name of your chip can be very heplful there.

GhidrAuto commented:
thank you for the quick response! I did what you suggested here https://github.com/NationalSecurityAgency/ghidra/issues/6863
